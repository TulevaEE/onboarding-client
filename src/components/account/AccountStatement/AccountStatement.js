import React from 'react';
import Types from 'prop-types';
import { FormattedMessage } from 'react-intl';

import Table from '../../common/table';
import Euro from '../../common/Euro';
import { getValueSum, getWeightedAverageFee } from './fundSelector';
import { Fees } from '../../common/Percentage/Fees';

const AccountStatement = ({ funds }) => {
  const valueSum = getValueSum(funds);
  const weightedAverageFee = getWeightedAverageFee(funds);

  const columns = [
    {
      title: <FormattedMessage id="accountStatement.columns.fund.title" />,
      dataIndex: 'fund',
      footer: <FormattedMessage id="accountStatement.columns.fund.footer" />,
    },
    {
      title: <FormattedMessage id="accountStatement.columns.fees.title" />,
      dataIndex: 'fees',
      footer: <Fees value={weightedAverageFee} />,
      hideOnMobile: true,
    },
    {
      title: <FormattedMessage id="accountStatement.columns.value.title" />,
      dataIndex: 'value',
      footer: <Euro amount={valueSum} />,
    },
  ];

  const dataSource = funds.map((fund) => {
    const className =
      !fund.activeFund && fund.price + fund.unavailablePrice === 0 ? 'text-muted' : undefined;
    return {
      fund: <span className={className}>{`${fund.name}${fund.activeFund ? '*' : ''}`}</span>,
      fees: <Fees className={className} value={fund.ongoingChargesFigure} />,
      value: <Euro className={className} amount={fund.price + fund.unavailablePrice} />,
      key: fund.isin,
    };
  });

  const showActiveFundNotice = funds.some(({ activeFund }) => activeFund);
  const fundPillar = funds.every(({ pillar }) => pillar === 2) ? 'secondPillar' : 'thirdPillar';

  return (
    <>
      <Table columns={columns} dataSource={dataSource} />

      {showActiveFundNotice && (
        <small className="text-muted">
          <FormattedMessage id={`accountStatement.${fundPillar}.activeFundNotice`} />
        </small>
      )}
    </>
  );
};

AccountStatement.propTypes = {
  funds: Types.arrayOf(
    Types.shape({
      isin: Types.string.isRequired,
      name: Types.string.isRequired,
      activeFund: Types.bool,
      contributions: Types.number,
      subtractions: Types.number,
      profit: Types.number,
      price: Types.number,
      unavailablePrice: Types.number,
    }),
  ).isRequired,
};

export default AccountStatement;
