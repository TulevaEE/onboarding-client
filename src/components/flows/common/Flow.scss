@import '../../variables';

.tv-step {
  position: relative;
}

.tv-steps > .tv-step:not(:first-child)::before {
  content: '';
  background-color: $gray-mid;
  display: inline-block;
  position: absolute;
  height: 25px;
  width: 1px;
  z-index: -1;
  left: 16px;
  top: -24px;
}
