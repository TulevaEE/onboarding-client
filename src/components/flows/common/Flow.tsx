import React from 'react';
import { Redirect, Route, Switch } from 'react-router-dom';
import { connect } from 'react-redux';

import StepTitle from './StepTitle';

import './Flow.scss';
import { State } from '../../../types';

export interface FlowStep {
  path: string;
  Component: () => React.ReactNode;
  title?: React.ReactNode;
}

export interface Props {
  steps: FlowStep[];
  flowPath: string;
  lastPartOfPath: string;
}

export const Flow: React.FC<Props> = ({ steps, flowPath, lastPartOfPath }) => {
  const stepPaths = steps.map((step) => step.path);
  const isStepPath = stepPaths.indexOf(lastPartOfPath) !== -1;
  const stepIndex = isStepPath ? stepPaths.indexOf(lastPartOfPath) : 0;
  const beforeSteps = steps.slice(0, stepIndex);
  const currentStep = steps[stepIndex];
  const afterSteps = steps.slice(stepIndex + 1);
  const currentStepPath = currentStep.path;
  const currentStepTitle = currentStep.title;
  const StepComponent = currentStep.Component;
  const isLastStep = stepIndex === steps.length - 1;

  return (
    <div className="row">
      <div className="col">
        <div className="tv-steps mt-4">
          <>
            {beforeSteps.map(({ path, title }, index) => (
              <StepTitle key={path} number={index + 1} completed>
                {title}
              </StepTitle>
            ))}
            {currentStepTitle && (
              <StepTitle number={stepIndex + 1} active>
                {currentStepTitle}
              </StepTitle>
            )}
          </>

          <div className="pb-5">
            <Switch>
              <Route path={`${flowPath}/${currentStepPath}`} render={StepComponent} />
              <Redirect exact path={`${flowPath}`} to={`${flowPath}/${currentStepPath}`} />
            </Switch>
          </div>
          {!isLastStep && <hr className="mb-4" />}
          {afterSteps.map(
            ({ path, title }, index) =>
              title && (
                <StepTitle key={path} number={index + 1 + stepIndex + 1}>
                  {title}
                </StepTitle>
              ),
          )}
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state: State) => ({
  lastPartOfPath: state.router.location.pathname.split('/').pop() || '',
});

export default connect(mapStateToProps)(Flow);
