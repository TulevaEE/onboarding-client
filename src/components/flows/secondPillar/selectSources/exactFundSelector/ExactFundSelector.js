/* eslint-disable react/no-array-index-key */
import React from 'react';
import { PropTypes as Types } from 'prop-types';
import { FormattedMessage } from 'react-intl';

import FundExchangeRow from './fundExchangeRow';
import { isTuleva } from '../../../../common/utils';

function createSelectionChangeHandler(index, selections, onChange) {
  return (newSelection) =>
    onChange([
      ...selections.slice(0, index),
      ...(newSelection ? [newSelection] : []),
      ...selections.slice(index + 1),
    ]);
}

function createRowAdder({ sourceFunds, targetFunds, selections, onChange }) {
  const tulevaFunds = targetFunds.filter((fund) => isTuleva(fund));
  return () =>
    onChange(
      selections.concat({
        sourceFundIsin: sourceFunds[0].isin,
        targetFundIsin:
          tulevaFunds[0].isin !== sourceFunds[0].isin ? tulevaFunds[0].isin : tulevaFunds[1].isin,
        percentage: 1,
      }),
    );
}

const ExactFundSelector = ({ selections, sourceFunds, targetFunds, onChange }) => (
  <>
    {selections && selections.length ? (
      <div className="row mt-3">
        {selections.map((selection, index) => (
          <FundExchangeRow
            key={index}
            selection={selection}
            sourceFunds={sourceFunds}
            targetFunds={targetFunds}
            onChange={createSelectionChangeHandler(index, selections, onChange)}
          />
        ))}
      </div>
    ) : (
      ''
    )}
    <button
      type="button"
      className="btn btn-outline-primary px-3 mt-3"
      onClick={createRowAdder({
        sourceFunds,
        targetFunds,
        selections,
        onChange,
      })}
    >
      <FormattedMessage id="select.sources.select.some.add" />
    </button>
  </>
);

const noop = () => null;

ExactFundSelector.defaultProps = {
  selections: [],
  sourceFunds: [],
  targetFunds: [],
  onChange: noop,
};

ExactFundSelector.propTypes = {
  selections: Types.arrayOf(Types.shape({})),
  sourceFunds: Types.arrayOf(Types.shape({})),
  targetFunds: Types.arrayOf(Types.shape({})),
  onChange: Types.func,
};

export default ExactFundSelector;
