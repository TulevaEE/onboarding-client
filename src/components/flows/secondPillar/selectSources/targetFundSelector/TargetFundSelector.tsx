import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl';

import './TargetFundSelector.scss';
import checkImage from '../../../common/SuccessNotice/success.svg';
import { Fund } from '../../../../common/apiModels';
import { Fees } from '../../../../common/Percentage/Fees';
import { TulevaFundIsin } from '../../../../common/utils';

type Props = {
  targetFunds: Fund[];
  onSelectFund: (fund: Fund) => void;
  isSelected: (fund: Fund) => boolean;
};

export const TargetFundSelector: React.FunctionComponent<Props> = ({
  targetFunds = [],
  onSelectFund,
  isSelected = () => false,
}) => {
  const { formatMessage } = useIntl();

  return (
    <div className="row mx-0 mt-3 tv-target-fund__container align-items-stretch">
      {targetFunds.map((fund) => (
        <div key={fund.isin} className="col-12 col-sm mb-2 me-2 p-0">
          <button
            type="button"
            className={`
                tv-target-fund h-100 p-3 text-start
                ${isSelected(fund) ? 'tv-target-fund--active' : ''}
              `}
            onClick={() => onSelectFund(fund)}
          >
            {isSelected(fund) ? (
              <div className="tv-target-fund__corner-check">
                <span>
                  <img src={checkImage} alt="Success" />
                </span>
              </div>
            ) : (
              ''
            )}
            <div className="tv-target-fund__inner-container small">
              <p className="mb-2 fs-6 fw-bold">{fund.name}</p>
              <p className="mb-3">
                <FormattedMessage
                  id={`target.funds.${fund.isin as TulevaFundIsin}.description`}
                  values={{
                    b: (chunks: string) => <b>{chunks}</b>,
                  }}
                />
              </p>
              <p className="m-0">
                <FormattedMessage id="target.funds.fees" />{' '}
                <Fees className="fw-bold" value={fund.ongoingChargesFigure} showPerYear />{' '}
                <span className="mx-1 text-separator">Â·</span>{' '}
                <a
                  className="tv-target-fund__terms-link"
                  href={formatMessage({
                    id: `target.funds.${fund.isin as TulevaFundIsin}.terms.link`,
                  })}
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <FormattedMessage id="target.funds.terms" />
                </a>
              </p>
            </div>
          </button>
        </div>
      ))}
    </div>
  );
};

export default TargetFundSelector;
