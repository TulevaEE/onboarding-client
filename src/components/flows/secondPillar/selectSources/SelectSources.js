import React, { useState } from 'react';
import { PropTypes as Types } from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { Link, Redirect } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';
import { selectExchangeSources, selectFutureContributionsFund } from '../../../exchange/actions';
import { ErrorMessage, Loader, Radio } from '../../../common';
import TargetFundSelector from './targetFundSelector';
import ExactFundSelector from './exactFundSelector';
import { isTuleva } from '../../../common/utils';
import AccountStatement from '../../../account/AccountStatement';
import { useMandateDeadlines } from '../../../common/apiHooks';
import { formatDateTime } from '../../../common/dateFormatter';

function selectAllWithTarget(sourceFunds, targetFund) {
  return sourceFunds
    .filter((fund, index, list) => list.length === 1 || fund.isin !== targetFund.isin)
    .map((fund) => ({
      sourceFundIsin: fund.isin,
      targetFundIsin: targetFund.isin,
      percentage: 1,
    }));
}

function selectionsError(selections) {
  const sourceFundPercentages = {};
  let errorDescriptionCode = null;
  selections.forEach((selection) => {
    if (!sourceFundPercentages[selection.sourceFundIsin]) {
      sourceFundPercentages[selection.sourceFundIsin] = 0;
    }
    sourceFundPercentages[selection.sourceFundIsin] += selection.percentage;
    if (sourceFundPercentages[selection.sourceFundIsin] > 1) {
      errorDescriptionCode = 'select.sources.error.source.fund.percentages.over.100';
    }

    if (selection.sourceFundIsin === selection.targetFundIsin) {
      errorDescriptionCode = 'select.sources.error.source.fund.is.target.fund';
    }
  });

  return errorDescriptionCode;
}

export const SelectSources = ({
  loadingSourceFunds,
  loadingTargetFunds,
  sourceFunds,
  targetFunds,
  sourceSelection,
  onSelectExchangeSources,
  onSelectFutureContributionsFund,
  sourceSelectionExact,
  error,
  nextPath,
  selectedFutureContributionsFundIsin,
}) => {
  const { data: mandateDeadlines } = useMandateDeadlines();
  const [someExistingSwitch, setSomeExistingSwitch] = useState(
    sourceSelection && sourceSelection.length > 0,
  );
  const [someFutureSwitch, setSomeFutureSwitch] = useState(
    selectedFutureContributionsFundIsin != null,
  );

  const onSomeExistingSwitchChange = (event) => {
    setSomeExistingSwitch(event.target.checked);
    onSelectExchangeSources([], true);
  };

  const onSomeFutureSwitchChange = (event) => {
    setSomeFutureSwitch(event.target.checked);
    onSelectFutureContributionsFund(null);
  };

  if (error) {
    return <ErrorMessage errors={error.body} />;
  }
  if (loadingSourceFunds || !sourceFunds || loadingTargetFunds || !targetFunds) {
    return <Loader className="align-middle" />;
  }
  const fullSelectionActive = !!sourceSelection.length && !sourceSelectionExact;
  const futureSelectionActive =
    !sourceSelection.length && !sourceSelectionExact && targetFunds && targetFunds.length;
  const validationErrorCode =
    sourceSelectionExact === true ? selectionsError(sourceSelection) : null;
  const isValid = !validationErrorCode;
  const tulevaTargetFunds =
    targetFunds && targetFunds.length && targetFunds.filter((fund) => isTuleva(fund));
  const defaultTargetFund =
    tulevaTargetFunds && tulevaTargetFunds.length ? tulevaTargetFunds[0] : null;

  function validationSelectionErrorElement(errorCode) {
    if (!errorCode) {
      return null;
    }
    return (
      <div className="text-danger mt-3">
        <FormattedMessage id={errorCode} />
      </div>
    );
  }

  return (
    <div>
      {sourceFunds && !sourceFunds.length && <Redirect to={nextPath} />}
      <div className="row">
        <div className="col-12">
          <div className="mb-4">
            <p>
              <FormattedMessage id="select.sources.intro" />
            </p>
            <p>
              <FormattedMessage id="select.sources.current.status" />
            </p>
            <AccountStatement
              funds={sourceFunds}
              activeFundNotice={<FormattedMessage id="select.sources.active.fund" />}
            />
          </div>
        </div>
      </div>
      <Radio
        name="tv-select-sources-type"
        id="tv-select-sources-type-all"
        selected={fullSelectionActive}
        onSelect={() =>
          onSelectExchangeSources(selectAllWithTarget(sourceFunds, defaultTargetFund), false)
        }
      >
        <h3 className="m-0">
          <FormattedMessage id="select.sources.select.all" />
        </h3>
        {fullSelectionActive ? (
          <div className="mt-3">
            <p>
              <FormattedMessage id="select.sources.select.all.subtitle" />
            </p>
            <p>
              {mandateDeadlines && (
                <FormattedMessage
                  id="select.sources.select.all.deadline"
                  values={{
                    periodEnding: formatDateTime(mandateDeadlines.periodEnding),
                    b: (chunks) => <b className="text-nowrap">{chunks}</b>,
                  }}
                />
              )}
            </p>
            <FormattedMessage id="select.sources.select.all.choose" />
            <TargetFundSelector
              targetFunds={tulevaTargetFunds}
              onSelectFund={(targetFund) =>
                onSelectExchangeSources(selectAllWithTarget(sourceFunds, targetFund), false)
              }
              selectedTargetFundIsin={sourceSelection[0].targetFundIsin}
            />
          </div>
        ) : (
          ''
        )}
      </Radio>
      <Radio
        name="tv-select-sources-type"
        id="tv-select-sources-type-future"
        className="mt-3"
        selected={futureSelectionActive}
        onSelect={() => onSelectExchangeSources([], false)}
      >
        <h3 className="m-0">
          <FormattedMessage id="select.sources.select.future" />
        </h3>
        {futureSelectionActive ? (
          <>
            <p className="mt-3">
              <FormattedMessage id="select.sources.select.future.subtitle" />
            </p>
            <p>
              <FormattedMessage id="select.sources.select.all.choose" />
            </p>
            <TargetFundSelector
              targetFunds={tulevaTargetFunds}
              onSelectFund={(fund) => onSelectFutureContributionsFund(fund.isin)}
              selectedTargetFundIsin={selectedFutureContributionsFundIsin}
            />
          </>
        ) : (
          ''
        )}
      </Radio>
      <Radio
        name="tv-select-sources-type"
        id="tv-select-sources-type-some"
        className="mt-3"
        selected={sourceSelectionExact}
        onSelect={() => onSelectExchangeSources(sourceSelection, true)}
      >
        <p className="m-0">
          <FormattedMessage id="select.sources.select.some" />
        </p>
        {sourceSelectionExact ? (
          <div className="mt-3">
            <hr className="mb-3" />

            <div className="custom-control custom-switch">
              <input
                type="checkbox"
                className="custom-control-input"
                id="someExistingSwitch"
                onChange={onSomeExistingSwitchChange}
                checked={someExistingSwitch}
              />
              <label className="custom-control-label text-bold" htmlFor="someExistingSwitch">
                <FormattedMessage id="select.sources.select.some.existing" />
              </label>
            </div>

            {someExistingSwitch ? (
              <>
                <ExactFundSelector
                  selections={sourceSelection}
                  sourceFunds={sourceFunds}
                  targetFunds={targetFunds}
                  onChange={(selection) => onSelectExchangeSources(selection, true)}
                />
                {validationSelectionErrorElement(validationErrorCode)}
              </>
            ) : (
              ''
            )}

            <hr className="mt-4 mb-3" />

            <div className="custom-control custom-switch">
              <input
                type="checkbox"
                className="custom-control-input"
                id="someFutureSwitch"
                onChange={onSomeFutureSwitchChange}
                checked={someFutureSwitch}
              />
              <label className="custom-control-label text-bold" htmlFor="someFutureSwitch">
                <FormattedMessage id="select.sources.select.some.future" />
              </label>
            </div>

            {someFutureSwitch ? (
              <>
                <select
                  className="custom-select mt-3"
                  onChange={(event) => onSelectFutureContributionsFund(event.target.value)}
                  value={selectedFutureContributionsFundIsin || ''}
                >
                  <FormattedMessage id="transfer.future.capital.other.fund">
                    {(message) => (
                      <option value="1" hidden="hidden">
                        {message}
                      </option>
                    )}
                  </FormattedMessage>
                  {targetFunds.map((fund) => (
                    <option value={fund.isin} key={fund.isin}>
                      {fund.name}
                    </option>
                  ))}
                </select>
              </>
            ) : (
              ''
            )}
          </div>
        ) : (
          ''
        )}
      </Radio>

      <Link id="nextStep" to={isValid ? nextPath : '#'}>
        <button type="button" className={`btn btn-primary mt-5 ${!isValid ? 'disabled' : ''}`}>
          <FormattedMessage id="steps.next" />
        </button>
      </Link>
    </div>
  );
};

const noop = () => null;

SelectSources.defaultProps = {
  sourceFunds: null,
  targetFunds: null,
  loadingSourceFunds: false,
  loadingTargetFunds: false,
  sourceSelection: [],
  sourceSelectionExact: false,
  onSelectExchangeSources: noop,
  onSelectFutureContributionsFund: noop,
  error: null,
  selectedFutureContributionsFundIsin: null,
};

SelectSources.propTypes = {
  sourceSelection: Types.arrayOf(Types.shape({ targetFundIsin: Types.string })),
  sourceSelectionExact: Types.bool,
  sourceFunds: Types.arrayOf(Types.shape({})),
  targetFunds: Types.arrayOf(Types.shape({})),
  loadingSourceFunds: Types.bool,
  loadingTargetFunds: Types.bool,
  onSelectExchangeSources: Types.func,
  onSelectFutureContributionsFund: Types.func,
  error: Types.shape({ body: Types.string }),
  nextPath: Types.string.isRequired,
  selectedFutureContributionsFundIsin: Types.string,
};

const mapStateToProps = (state) => ({
  sourceSelection: state.exchange.sourceSelection,
  sourceSelectionExact: state.exchange.sourceSelectionExact,
  sourceFunds: state.exchange.sourceFunds,
  targetFunds: state.exchange.targetFunds,
  loadingSourceFunds: state.exchange.loadingSourceFunds,
  loadingTargetFunds: state.exchange.loadingTargetFunds,
  error: state.exchange.error,
  selectedFutureContributionsFundIsin: state.exchange.selectedFutureContributionsFundIsin,
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      onSelectExchangeSources: selectExchangeSources,
      onSelectFutureContributionsFund: selectFutureContributionsFund,
    },
    dispatch,
  );

const connectToRedux = connect(mapStateToProps, mapDispatchToProps);

export default connectToRedux(SelectSources);
