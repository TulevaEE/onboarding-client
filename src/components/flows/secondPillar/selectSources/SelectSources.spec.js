import React from 'react';
import { shallow } from 'enzyme';

import { Redirect } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';
import { ErrorMessage, Loader, Radio } from '../../../common';
import ExactFundSelector from './exactFundSelector';
import TargetFundSelector from './targetFundSelector';
import { SelectSources } from './SelectSources';
import AccountStatement from '../../../account/AccountStatement';

// TODO: Figure out a cleaner way to mock the hook
jest.mock('../../../common/apiHooks', () => ({
  useMandateDeadlines: () => ({ data: { periodEnding: '2023-07-31T20:59:59.999999999Z' } }),
}));

describe('Select sources step', () => {
  let component;

  beforeEach(() => {
    component = shallow(
      <SelectSources nextPath="/2nd-pillar-flow/address" sourceFunds={[]} targetFunds={[]} />,
    );
  });

  it('renders a loader when loading source or target funds', () => {
    component.setProps({ loadingSourceFunds: true });
    expect(component.get(0)).toEqual(<Loader className="align-middle" />);
    component.setProps({ loadingSourceFunds: false, loadingTargetFunds: true });
    expect(component.get(0)).toEqual(<Loader className="align-middle" />);
  });

  it('does not render a loader when funds loaded', () => {
    component.setProps({
      loadingSourceFunds: false,
      loadingTargetFunds: false,
      sourceFunds: [{ isin: 'a' }],
      targetFunds: [{ isin: 'b' }],
    });
    expect(component.get(0)).not.toEqual(<Loader className="align-middle" />);
  });

  it('renders a pension fund table with given funds', () => {
    const sourceFunds = [{ iAmAFund: true }, { iAmAlsoAFund: true }];
    component.setProps({ sourceFunds, targetFunds: [] });
    expect(
      component.contains(
        <AccountStatement
          funds={sourceFunds}
          activeFundNotice={<FormattedMessage id="select.sources.active.fund" />}
        />,
      ),
    ).toBe(true);
  });

  it('renders a button to the next step', () => {
    const nextPath = '/next-path';
    component.setProps({ nextPath });

    expect(component.find('#nextStep').prop('to')).toBe(nextPath);
    expect(component.find('#nextStep button').children().at(0).getElement(0)).toMatchSnapshot();
  });

  it('on exact selection disables the next step button if selection is invalid', () => {
    component.setProps({
      sourceSelection: [{ sourceFundIsin: 'a', percentage: 1 }],
      selectedFutureContributionsFundIsin: 'a',
      onSelectExchangeSources: jest.fn(),
    });

    component.find(Radio).last().simulate('select');
    component.find('#someExistingSwitch').simulate('change', { target: { checked: true } });

    expect(component.find('#nextStep button').prop('className')).not.toContain('disabled');
    expect(
      component.contains(
        <FormattedMessage id="select.sources.error.source.fund.percentages.over.100" />,
      ),
    ).toBe(false);

    component.setProps({
      sourceSelection: [
        { sourceFundIsin: 'a', percentage: 1 },
        { sourceFundIsin: 'a', percentage: 1 },
      ],
      selectedFutureContributionsFundIsin: 'a',
    });
    expect(component.find('#nextStep button').prop('className')).toContain('disabled');
    expect(
      component.contains(
        <FormattedMessage id="select.sources.error.source.fund.percentages.over.100" />,
      ),
    ).toBe(true);
  });

  it('on exact selection disables the next step button if selection is invalid due to inter fund transfer and displays help text', () => {
    component.setProps({
      sourceSelection: [
        {
          sourceFundIsin: 'a',
          targetFundIsin: 'b',
          percentage: 1,
        },
      ],
      onSelectExchangeSources: jest.fn(),
    });
    component.find(Radio).last().simulate('select');
    component.find('#someExistingSwitch').simulate('change', { target: { checked: true } });

    expect(component.find('#nextStep button').prop('className')).not.toContain('disabled');
    expect(
      component.contains(<FormattedMessage id="select.sources.error.source.fund.is.target.fund" />),
    ).toBe(false);

    component.setProps({
      sourceSelection: [
        {
          sourceFundIsin: 'a',
          targetFundIsin: 'a',
          percentage: 1,
        },
      ],
    });
    expect(component.find('#nextStep button').prop('className')).toContain('disabled');

    expect(
      component.contains(<FormattedMessage id="select.sources.error.source.fund.is.target.fund" />),
    ).toBe(true);
  });

  it('sets the full selection radio by default', () => {
    const fullSelectionRadio = () => component.find(Radio).first();
    component.setProps({
      sourceSelection: [],
    });
    expect(fullSelectionRadio().prop('selected')).toBe(true);

    component.find(Radio).last().simulate('select');
    expect(fullSelectionRadio().prop('selected')).toBe(false);
  });

  it('selects all funds when toggling the someExistingSwitch on', () => {
    const onSelectExchangeSources = jest.fn();
    const sourceFunds = [{ isin: 'a' }, { isin: 'b' }];
    const targetFunds = [{ isin: 'c', fundManager: { name: 'Tuleva' } }];
    const fullSelection = [
      { sourceFundIsin: 'a', targetFundIsin: 'c', percentage: 1 },
      { sourceFundIsin: 'b', targetFundIsin: 'c', percentage: 1 },
    ];
    component.setProps({ sourceFunds, targetFunds, onSelectExchangeSources });
    expect(onSelectExchangeSources).not.toHaveBeenCalled();
    component.find(Radio).last().simulate('select');
    component.find('#someExistingSwitch').simulate('change', { target: { checked: true } });
    expect(onSelectExchangeSources).toHaveBeenCalledTimes(1);
    expect(onSelectExchangeSources).toHaveBeenCalledWith(fullSelection, true);
  });

  it('when selecting all funds, skip inter fund transfer', () => {
    const onSelectExchangeSources = jest.fn();
    const sourceFunds = [{ isin: 'c', fundManager: { name: 'Tuleva' } }, { isin: 'b' }];
    const targetFunds = [{ isin: 'c', fundManager: { name: 'Tuleva' } }];
    const fullSelection = [
      {
        sourceFundIsin: 'b',
        targetFundIsin: 'c',
        percentage: 1,
      },
    ];
    component.setProps({ sourceFunds, targetFunds, onSelectExchangeSources });
    expect(onSelectExchangeSources).not.toHaveBeenCalled();
    component.find(Radio).last().simulate('select');
    component.find('#someExistingSwitch').simulate('change', { target: { checked: true } });
    expect(onSelectExchangeSources).toHaveBeenCalledTimes(1);
    expect(onSelectExchangeSources).toHaveBeenCalledWith(fullSelection, true);
  });

  it('when selecting all funds and the default fund (first target fund) is the only source fund, dont skip it', () => {
    const onSelectExchangeSources = jest.fn();
    const sourceFunds = [{ isin: 'c', fundManager: { name: 'Tuleva' } }];
    const targetFunds = [{ isin: 'c', fundManager: { name: 'Tuleva' } }, { isin: 'b' }];
    const fullSelection = [
      {
        sourceFundIsin: 'c',
        targetFundIsin: 'c',
        percentage: 1,
      },
    ];
    component.setProps({ sourceFunds, targetFunds, onSelectExchangeSources });
    expect(onSelectExchangeSources).not.toHaveBeenCalled();
    component.find(Radio).last().simulate('select');
    component.find('#someExistingSwitch').simulate('change', { target: { checked: true } });
    expect(onSelectExchangeSources).toHaveBeenCalledTimes(1);
    expect(onSelectExchangeSources).toHaveBeenCalledWith(fullSelection, true);
  });

  it('sets the exact fund selector active as selected when fund selection is exact', () => {
    const exactFundSelector = () => component.find(Radio).last();
    expect(exactFundSelector().prop('selected')).toBe(false);
    component.find(Radio).last().simulate('select');
    expect(exactFundSelector().prop('selected')).toBe(true);
  });

  it('shows the exact fund selector when fund selection is exact', () => {
    const exactFundSelectorRendered = () => !!component.find(ExactFundSelector).length;
    component.setProps({ onSelectExchangeSources: jest.fn() });
    component.find(Radio).last().simulate('select');
    component.find('#someExistingSwitch').simulate('change', { target: { checked: true } });
    expect(exactFundSelectorRendered()).toBe(true);
    component.find('#someExistingSwitch').simulate('change', { target: { checked: false } });
    expect(exactFundSelectorRendered()).toBe(false);
  });

  it('shows switches for exact fund selection when selection is exact', () => {
    const subtitleRendered = () =>
      component.contains(<FormattedMessage id="select.sources.select.some.existing" />) &&
      component.contains(<FormattedMessage id="select.sources.select.some.future" />);
    expect(subtitleRendered()).toBe(false);
    component.find(Radio).last().simulate('select');
    expect(subtitleRendered()).toBe(true);
  });

  it('passes the current selection and funds to the exact fund selector', () => {
    const sourceSelection = [
      { sourceFundIsin: 'a', targetFundIsin: 'c', percentage: 1 },
      { sourceFundIsin: 'b', targetFundIsin: 'c', percentage: 1 },
    ];
    const sourceFunds = [{ isin: 'a' }, { isin: 'b' }];
    const targetFunds = [{ isin: 'c', fundManager: { name: 'Tuleva' } }];
    component.setProps({
      sourceSelection,
      sourceFunds,
      targetFunds,
      onSelectExchangeSources: jest.fn(),
    });
    component.find(Radio).last().simulate('select');
    component.find('#someExistingSwitch').simulate('change', { target: { checked: true } });

    const selectorProp = (name) => component.find(ExactFundSelector).prop(name);
    expect(selectorProp('selections')).toBe(sourceSelection);
    expect(selectorProp('sourceFunds')).toBe(sourceFunds);
    expect(selectorProp('targetFunds')).toBe(targetFunds);
  });

  it('selects exact funds when the selector tells it to', () => {
    const onSelectExchangeSources = jest.fn();
    component.setProps({ onSelectExchangeSources });
    const selection = [
      { name: 'a', percentage: 0.7 },
      { name: 'b', percentage: 0.8 },
    ];
    expect(onSelectExchangeSources).not.toHaveBeenCalled();
    component.find(Radio).last().simulate('select');
    component.find('#someExistingSwitch').simulate('change', { target: { checked: true } });
    component.find(ExactFundSelector).simulate('change', selection);
    expect(onSelectExchangeSources).toHaveBeenCalledTimes(2);
    expect(onSelectExchangeSources).toHaveBeenCalledWith(selection, true);
  });

  it('passes only Tuleva funds to the target fund selector', () => {
    const tulevaFund = { isin: 'b', fundManager: { name: 'Tuleva' } };
    const targetFunds = [{ isin: 'a' }, tulevaFund];

    component.setProps({
      sourceSelection: [{ name: 'a', percentage: 1 }],
      targetFunds,
    });
    expect(component.find(TargetFundSelector).prop('targetFunds')).toEqual([tulevaFund]);
  });

  it('renders error', () => {
    const error = { body: 'aww no' };
    const funds = [{ aFund: true }];

    component.setProps({ error, funds });

    expect(component.contains(<ErrorMessage errors={error.body} />)).toBe(true);
    expect(component.contains(<Loader className="align-middle" />)).toBe(false);
    expect(component.contains(<AccountStatement funds={funds} />)).toBe(false);
  });
});
