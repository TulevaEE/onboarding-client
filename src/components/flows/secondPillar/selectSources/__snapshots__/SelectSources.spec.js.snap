// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select sources step renders a button to the next step 1`] = `
<FormattedMessage
  id="steps.next"
/>
`;
